### ISSUE TITLE
bug: <dialog> element items are focusable, even when <dialog> element is closed

### ISSUE THREAD

**Issue by viktorb1** [saadeghi/daisyui#2223]
### What version of daisyUI are you using?

3.5.1

### Describe your issue

When you use a dialog modal such as https://daisyui.com/components/modal/#dialog-modal, everything is still tabbable even when the modal is not open. The problem seems to be caused by display: grid preventing the default behavior of the <dialog> element from being set to display: none.

### What browsers are you seeing the problem on?

Chrome

### Reproduction URL

[Your website demos, for example](https://daisyui.com/components/modal/#dialog-modal)

**Comment by saadeghi** [comment]
Thank you for reporting.  
Unfortunately removing the `grid` (and adding it only when modal is open) disables the open animation.  
I don't know if there's a way to fix that yet, but I don't want fixing one issue to create another.

**Comment by viktorb1** [comment]
In my project, I dynamically add/remove the 'modal' class when the <dialog> element opens/closes and it works as expected. Decent workaround for the time being if anyone else runs into this problem.

**Comment by dogagenc** [comment]
@saadeghi This is being a problem both for general keyboard usage and accessibility. Is there a reason you don't want to use `visibility: hidden` and `visibility: visible` for close/open states? - which seems it's being used in `unstyled` components - 

It shouldn't affect animation.

**Comment by velut** [comment]
A quick solution (sacrificing the open animation) to restore the default `display:none` behavior seems to be using `open:modal` instead of `modal` as the dialog class.

```html
<dialog id="my_modal_1" class="open:modal">
  <!-- Change the class here   ^^^^^^^^^^ -->
</dialog>
```

**Comment by cknight** [comment]
> In my project, I dynamically add/remove the 'modal' class when the element opens/closes and it works as expected. Decent workaround for the time being if anyone else runs into this problem.

I've just run into this and your workaround is good, but doesn't work if the user presses escape to close the dialog.  In this case, the dialog closes, but still contains the `modal` class.  Did you find a workaround for that?

**Comment by viktorb1** [comment]
@cknight It's been a while since I've had this issue, but what is stopping you from writing a keypress event handler in javascript?

**Comment by oyvindlille** [comment]
Any progress on this? This is in conflict of multiple wcag critera like 2.4.3.
My fix was to use open:modal, but suggest this to be fixed on the class .modal or at least be mentioned in the documentation.

**Comment by Atmos4** [comment]
Achieving this in pure CSS is from my experience very hard. 
However, there is a new `transition-behavior` option for CSS transitions called `allow-discrete`. It basically allows to play the closing animation for native `dialog` elements before hiding them.

Given that DaisyUI is supposed to have good browser support I don't think this solution is good enough. The solution I have thus been using with DaisyUI is to add `visibility:hidden` manually after a delay when the modal is being closed. This can be for example achieved by listening to the `transitionend` event, or some other more manual style manipulation with Javascript.

**Comment by YouSafe** [comment]
I ran into this too when I noticed that on Android Chrome the chess pieces that are behind my modal's textarea can not be dragged. I'm using lichess' [chessground](https://github.com/lichess-org/chessground) chess board, which checks if the TouchEvent is `cancelable` ([see](https://github.com/lichess-org/chessground/blob/fad0543fcbd3662e7af16a14af04e1993f941a89/src/drag.ts#L41)) and for the area behind the textarea the value in the event is `false` and everywhere else on the board `true`. 

Adding the `invisible` class to the closed dialog has resolved this issue for me.

**Comment by Nikhil1920** [comment]
There should be a note in the docs regarding this

Having better keyboard accessibility might be more important than opening animation for some people

**Comment by acaloiaro** [comment]
An additional side effect of `display: grid` on the `modal` class is that intersection observance (https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) on elements within `<dialog class="modal">` is not possible because everything within the modal will have intersected the viewport immediately upon page load.

In other words, it's not possible to lazy load content in models if the content will be visible when the modal is visible.

**Comment by acaloiaro** [comment]
@Atmos4 could you clarify what about my comment you're thumbing down? 

I was under the impression that I was making an objective comment about an observation. If there's something about it that's wrong, I'd appreciate a correction.

**Comment by Atmos4** [comment]
That is true, apologies. The reason I thumbed down is because I do not believe an intersection observer is the right way to load the content of a modal. If you wish to lazy-load a modal content, you should do either when it turns visible or right after it is loaded on the page.

Trying to use intersection observers with modals is generally a bad idea and you will face the problem you are having with many modal implementations out there 🙂

**Comment by acaloiaro** [comment]
Thanks for the clarification. 

I don't fully agree with your opinion, although I'll concede there's a chance I'm in the minority. I'm curious to see where this thread lands on the topic. 

With that said, the `<dialog>` element, the browser's native modal element, has `display: none` in the unopened state (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog#animating_dialogs). That means using intersection observance is a great way to lazy load content within dialogs. You did say modals more generally, and I'm specifically talking about `<dialog>`. I'd agree that using intersection observance in conjunction with many CSS packages' `modal` class is probably not a good idea because every package has different priorities, and thus implementation details to consider. That doesn't make intersection observance within dialogs a bad idea on its own. As daisyUI has obviated in this instance -- some packages are willing to prioritize package-specific goals like animations over the problems some of us have faced with daisyUI's modal `grid` display decision. That's what makes using intersection observance a potentially bad idea in daisyUI `modal`s. 

I'll argue along with @Nikhil1920 that as long as daisyUI prioritizes animations over other things, a lot of developers will appreciate a caveat in the `modal` docs. DaisyUI/maintainers are entitled to prioritize whatever achieves the best outcome for them, I'll only argue that the footgun should be called out in the docs.

**Comment by ddzwigaj** [comment]
This just crashed my website on mobile, every time user enters the website the input inside dialog is auto focused because its type email and safari tries to auto login user

**Comment by saadeghi** [comment]
@ddzwigaj does it "crash?" 🤔  

Issue is fixed in daisyUI 5 beta
